### PR TITLE
Dataplane telemetry support in SONiC

### PR DESCRIPTION
**Pull request by shruthi9**
This includes changes to make SONiC work with SAIv 1.3, as DTel API support is only in v 1.3. These changes can be reverted when official SONiC support for v1.3 is added.

Design for the feature is at:
https://github.com/Azure/SONiC/pull/182/

### CODE REVIEW

**Comment by lguohan** [comment]
@prsunny , can you take a look at this PR and check if you are missing anything in your PR?

File: syncd/Makefile.am [hunk 1]
```diff
@@ -29,6 +34,7 @@ syncd_LDADD = -lhiredis -lswsscommon $(SAILIB) -lpthread -L$(top_srcdir)/meta/.l
 if SAITHRIFT
 syncd_CPPFLAGS += -DSAITHRIFT=yes
 syncd_LDADD += -lrpcserver -lthrift
+#syncd_LDADD += -lthrift
```
Comment (lguohan) [inline syncd/Makefile.am]: remove this commented line.

File: saidiscovery/Makefile.am [hunk 2]
```diff
@@ -11,8 +11,12 @@ endif
 if SAIVS
 SAILIB=-L$(top_srcdir)/vslib/src/.libs -lsaivs
 else
+if sonic_asic_platform_barefoot
+SAILIB=-L/opt/bfn/install/lib -lswitchsai -lswitchapi -lbf_switchd_lib -ltofinopdfixed_thrift -ldriver -lbfutils -lbfsys -lbfutils -L/opt/bfn/install/lib/tofinopd/switch -lpd -lpdcli -lpdthrift
+else
```
Comment (lguohan) [inline saidiscovery/Makefile.am]: remove -lsai.

**Comment by lguohan** [comment]
Can you resolve the conflict?

File: lib/src/sai_redis_bridge.cpp [hunk 3]
```diff
@@ -13,10 +13,12 @@ sai_status_t redis_get_bridge_stats(
     return SAI_STATUS_NOT_IMPLEMENTED;
 }
 
-sai_status_t redis_clear_bridge_stats(
+sai_status_t redis_get_bridge_stats_ext(
```
Comment (lguohan) [inline lib/src/sai_redis_bridge.cpp]: i am not sure this needs to be modified.

Comment (mkbalani) [reply lib/src/sai_redis_bridge.cpp]: fixed

File: vslib/src/sai_vs_dtel.cpp [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+VS_GENERIC_QUAD(DTEL,dtel);
+VS_GENERIC_QUAD(DTEL_QUEUE_REPORT,dtel_queue_report);
+VS_GENERIC_QUAD(DTEL_INT_SESSION,dtel_int_session);
+VS_GENERIC_QUAD(DTEL_REPORT_SESSION,dtel_report_session);
+VS_GENERIC_QUAD(DTEL_EVENT,dtel_event);
+
+const sai_dtel_api_t vs_dtel_api = {
+
+	VS_GENERIC_QUAD_API(dtel)
```
Comment (lguohan) [inline vslib/src/sai_vs_dtel.cpp]: can you convert all tabs to spaces?

File: syncd/Makefile.am [hunk 5]
```diff
@@ -11,8 +11,13 @@ endif
 if SAIVS
 SAILIB=-L$(top_srcdir)/vslib/src/.libs -lsaivs
 else
+if sonic_asic_platform_barefoot
+AM_CPPFLAGS += -I/opt/bfn/install/include
+SAILIB=-L/opt/bfn/install/lib -lswitchsai -lswitchapi -lbf_switchd_lib -ltofinopdfixed_thrift -ldriver -lbfutils -lbfsys -lbfutils -L/opt/bfn/install/lib/tofinopd/switch -lpd -lpdcli -lpdthrift
```
Comment (lguohan) [inline syncd/Makefile.am]: only link to libsai

File: lib/src/sai_redis_dtel.cpp [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+REDIS_GENERIC_QUAD(DTEL,dtel);
+REDIS_GENERIC_QUAD(DTEL_QUEUE_REPORT,dtel_queue_report);
+REDIS_GENERIC_QUAD(DTEL_INT_SESSION,dtel_int_session);
+REDIS_GENERIC_QUAD(DTEL_REPORT_SESSION,dtel_report_session);
+REDIS_GENERIC_QUAD(DTEL_EVENT,dtel_event);
+
+const sai_dtel_api_t redis_dtel_api = {
+
+	REDIS_GENERIC_QUAD_API(dtel)
```
Comment (lguohan) [inline lib/src/sai_redis_dtel.cpp]: TAB -> SPACES

File: saisdkdump/Makefile.am [hunk 7]
```diff
@@ -8,11 +8,16 @@ else
 DBGFLAGS = -g
 endif
 
+
 if SAIVS
 SAILIB=-L$(top_srcdir)/vslib/src/.libs -lsaivs
 else
+if sonic_asic_platform_barefoot
+SAILIB=-L/opt/bfn/install/lib -lswitchsai -lswitchapi -lbf_switchd_lib -ltofinopdfixed_thrift -ldriver -lbfutils -lbfsys -lbfutils -L/opt/bfn/install/lib/tofinopd/switch -lpd -lpdcli -lpdthrift
```
Comment (lguohan) [inline saisdkdump/Makefile.am]: remove library other than the sai package.

File: syncd/scripts/syncd_init_common.sh [hunk 8]
```diff
@@ -103,10 +103,28 @@ config_syncd_marvell()
...
+    P4_PROFILE=$(sonic-cfggen -d -v 'DEVICE_METADATA["localhost"]["p4_profile"]')
+    if [[ -n "$P4_PROFILE" ]]; then
+        if [[ ( -d /opt/bfn/install_${P4_PROFILE} ) && ( -L /opt/bfn/install || ! -e /opt/bfn/install ) ]]; then
+            ln -srfn /opt/bfn/install_${P4_PROFILE} /opt/bfn/install
+        fi
+    fi
+    export ONIE_PLATFORM=`grep onie_platform /etc/machine.conf | awk 'BEGIN { FS = "=" } ; { print $2 }'`
+    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/bfn/install/lib/platform/$ONIE_PLATFORM:/opt/bfn/install/lib:/opt/bfn/install/lib/tofinopd/switch
+     . /opt/bfn/install/bin/dma_setup.sh
+     export LD_PRELOAD=libswitchapi.so:libswitchsai.so:libpd.so:libpdcli.so:libdriver.so:libbfsys.so:libbfutils.so:libbf_switchd_lib.so:libtofinopdfixed_thrift.so:libpdthrift.so
```
Comment (lguohan) [inline syncd/scripts/syncd_init_common.sh]: alignment here is not consistent.

File: vslib/src/sai_vs_policer.cpp [hunk 9]
```diff
@@ -40,6 +40,20 @@ sai_status_t vs_clear_policer_stats(
     return SAI_STATUS_NOT_IMPLEMENTED;
 }
 
+sai_status_t vs_get_policer_stats_ext(
```
Comment (lguohan) [inline vslib/src/sai_vs_policer.cpp]: this one should replace vs_sai_get_policer_stats_ext

Please remove vs_sai_get_policer_stats_ext above.

File: saisdkdump/Makefile.am (hunk 7 above)
Comment (mkbalani) [reply saisdkdump/Makefile.am]: fixed

File: syncd/scripts/syncd_init_common.sh (hunk 8 above)
Comment (mkbalani) [reply syncd/scripts/syncd_init_common.sh]: fixed

File: syncd/Makefile.am (hunk 5 above)
Comment (mkbalani) [reply syncd/Makefile.am]: fixed

**Pull request merged**